src/Npgsql/ConnectorPool.cs

```diff
@@ -413,117 +379,73 @@ internal void Release(NpgsqlConnector connector)
-                        }
-                    }
-                    sw.SpinOnce();
-                }
+            // Scenario: pre-empted release
+            // Right here between our check for waiters and our signalling decrement for storing
+            // a connector there could have been a new waiter enqueueing, we compensate at the end.
+
+            // If we're here, we put the connector back in the idle list
+            // We increment Idle, any allocate that is racing us will not match Busy == _max
+            // and will not enqueue but try to get our connector.
+            Interlocked.Increment(ref State.Idle);
+            connector.ReleaseTimestamp = DateTime.UtcNow;
+            // Prevent reordering wit ReleaseTimestamp
+            Volatile.Write(ref _idle[connector.PoolIndex], connector);
+            CheckInvariants(State);
+
+            // Scenario: pre-empted release
+            // We checked at the start of release if there were any waiters.
+            // Unblock any new waiter that raced us by handing it a null result.
+            // We try to complete exactly one waiter as long as there are any in the queue, if any came in at all.
+            // The performance of trying this after each _idle release is fine as the queue is very uncontended.
+            // In the .Net Core BCL, 3.0 as of writing, TryDequeue for the empty path is as fast as doing IsEmpty.
+            while (_waiting.TryDequeue(out var racedWaiter))
+            {
+                if (racedWaiter.TaskCompletionSource.TrySetResult(null))
+                    break;
             }
+
+            // Scenario: pre-empted waiter
+            // Could have a pre-empted waiter, that didn't enqueue yet it wakes up right after
+            // our correcting dequeue, it will do its own check after that Enqueue for Idle > 0.
         }
 
         void CloseConnector(NpgsqlConnector connector, bool wasIdle)
         {
             try
             {
                 connector.Close();
-
-                var sw = new SpinWait();
-                while (true)
-                {
-                    var state = State.Copy();
-                    var newState = state;
-                    if (wasIdle)
-                        newState.Idle--;
-                    else
-                        newState.Busy--;
-                    CheckInvariants(newState);
-                    if (Interlocked.CompareExchange(ref State.All, newState.All, state.All) == state.All)
-                        break;
-                    sw.SpinOnce();
-                }
             }
             catch (Exception e)
             {
                 Log.Warn("Exception while closing outdated connector", e, connector.Id);
             }
 
+            _open[connector.PoolIndex] = null;
+
+            // We decrement open first so we don't race waiters that could otherwise see Busy == _max and enqueue.
+            Interlocked.Decrement(ref State.Open);
```

Comment:
As above, I'd prefer to see this in a CAS loop, just to avoid having to reason about why it's OK... The cost here is completely negligible (nobody cares about physical close).

Reply:
Will change!